go/vt/vtgate/planbuilder/builder.go

```diff
@@ -151,9 +152,83 @@ func buildRoutePlan(stmt sqlparser.Statement, reservedVars *sqlparser.ReservedVa
 	return f(stmt, reservedVars, vschema)
 }
 
+func buildRoutePlanWithMirroring(stmt sqlparser.Statement, reservedVars *sqlparser.ReservedVars, vschema plancontext.VSchema, f func(statement sqlparser.Statement, reservedVars *sqlparser.ReservedVars, schema plancontext.VSchema) (*planResult, error)) (*planResult, error) {
+	// Building a plan changes the statement, so clone it before we create the
+	// original plan.
+	//
+	// TODO(maxeng): find a way to avoid this work unless all tables have a
```

Comment:
just to clarify, `CloneStatement` is required in order to mirror. we are calling `buildRoutePlan` twice: once for the original plan, and once for the mirror target. running `buildRoutePlan` modifies the input `stmt`, so that the same `stmt` can't be used as input to multiple `buildRoutePlan`.

however, ideally i wouldn't call `CloneStatement` until _after_ i know mirroring is called for. the goal of the `TODO` is to figure out how to do that.

i think the problem with your suggestion is that semantic state isn't populated until `semantics.Analyze` is called, which isn't called until `planningcontext.CreatePlanningContext` is called, and so forth until `buildRoutePlan` is called.

if i run `buildRoutePlan` in order to obtain the semantic state, then i can't re-use that same `stmt` input into that function to again to build the mirror plan. which means i need to `CloneStatement` first.

i could try to hoist `planningcontext.CreatePlanningContext` to higher up in the call stack, but that actually looks quite difficult. for example in `SELECT` planning, in some cases `stmt` is rewritten, and the result of the rewrite is passed in to `CreatePlanningContext`, which means i would have to hoist up that rewrite logic as well.

Reply:
a couple other ideas:

 * i could just add a brand new `semantic.Analyze` call inside of `buildRoutePlanWithMirroring`, and use that to determine whether mirroring is called for. but that will also add unwanted overhead, similar to `CloneStatement`, and it won't eliminate the need for `CloneStatement`.
 * we can simply short-circuit mirror planning if there are zero mirror rules defined in the VSchema. mirror rules are only created with (optional) `MoveTables MirrorTraffic`, and are always removed with `MoveTables SwitchTraffic` (required). so only callers of `MoveTables MirrorTraffic` would be punished with the `CloneStatement` overhead, and only until `SwitchTraffic` removes the rules.

Reply:
To add a bit more context to this:

> running buildRoutePlan modifies the input stmt, so that the same stmt can't be used as input to multiple buildRoutePlan.

I'm not sure how many modifications are happening that make this true. But the main one seems to be this:

https://github.com/vitessio/vitess/blob/3d36adb16eff96485db992454e2e287545367035/go/vt/vtgate/planbuilder/single_sharded_shortcut.go#L33-L43

If I comment that out, I'm able to re-use the same input `stmt` into both the original `buildRoutePlan` and mirror `buildRoutePlan` calls. Of course that may result in invalid plans, but the point is that if we're open  to finding other ways of planning the unsharded shortcut that doesn't result in modifying the input `stmt`, that may let us circumvent needing to `CloneStatement`.

Reply:
i made an attempt to do this in [2cd6230](https://github.com/vitessio/vitess/pull/15992/commits/2cd623065469d9d2da1f900e769c58236cbb028d)

> if we're open to finding other ways of planning the unsharded shortcut that doesn't result in modifying the input stmt, that may let us circumvent needing to CloneStatement.

instead of modifying the `SELECT` statement, i used a `sqlparser.NodeFormatter` to selectively clone-and-modify the parts of the statement that need the keyspace qualifier to be removed. this allows me to re-use the same input `stmt` into multiple `buildRoutePlan` calls without having to first `CloneStatement`.

while this works fine for the handful of cases currently in `mirror_cases.json`, there may be other instances where the `SELECT` statement is modified in-place that would cause the mirrored statement to be incorrect. but, if you are happy with this approach, i can add more test cases to make sure we're well covered.

Reply:
based on my discussion with you and Harshit, in c8ada40b13f2c19b018e6ea775034dbf92735df4:

 * use `semantics` `PlanningContext` and `SemTable` to keep track of whether mirroring is required.
 * moved of the mirror planning logic from `go/vt/vtgate/planbuilder/builder.go` into `go/vt/vtgate/planbuilder/operators`.